src/trusted/service_runtime/sel_ldr_standard.c

```diff
@@ -904,6 +967,214 @@ int NaClCreateMainThread(struct NaClApp     *nap,
+                             char const *const        *envv) {
+  /*
+   * Compute size of string tables for argv and envv
+   */
+  int                   retval;
+  int                   envc;
+  size_t                size;
+  int                   auxv_entries;
+  size_t                ptr_tbl_size;
+  int                   i;
+  uint32_t              *p;
+  char                  *strp;
+  size_t                *argv_len;
+  size_t                *envv_len;
+  uintptr_t             stack_ptr;
+
+  retval = 0;  /* fail */
+  CHECK(argc >= 0);
+  CHECK(argv || argc);
+
+  envc = 0;
+  if (NULL != envv) {
+    char const *const *pp;
+    for (pp = envv; NULL != *pp; ++pp) {
+      ++envc;
+    }
+  }
+  envv_len = 0;
+  argv_len = !argc ? NULL : malloc(argc * sizeof argv_len[0]);
+  envv_len = !envc ? NULL : malloc(envc * sizeof envv_len[0]);
+  if (NULL == argv_len) {
+    goto cleanup;
+  }
+  if (NULL == envv_len && 0 != envc) {
+    goto cleanup;
+  }
+
+  size = 0;
+
+  /*
+   * The following two loops cannot overflow.  The reason for this is
+   * that they are counting the number of bytes used to hold the
+   * NUL-terminated strings that comprise the argv and envv tables.
+   * If the entire address space consisted of just those strings, then
+   * the size variable would overflow; however, since there's the code
+   * space required to hold the code below (and we are not targetting
+   * Harvard architecture machines), at least one page holds code, not
+   * data.  We are assuming that the caller is non-adversarial and the
+   * code does not look like string data....
+   */
+  for (i = 0; i < argc && argv; ++i) {
+    argv_len[i] = strlen(argv[i]) + 1;
+    size += argv_len[i];
+  }
+  for (i = 0; i < envc; ++i) {
+    envv_len[i] = strlen(envv[i]) + 1;
+    size += envv_len[i];
+  }
+
+
+  /*
+   * NaCl modules are ILP32, so the argv, envv pointers, as well as
+   * the terminating NULL pointers at the end of the argv/envv tables,
+   * are 32-bit values.  We also have the auxv to take into account.
+   *
+   * The argv and envv pointer tables came from trusted code and is
+   * part of memory.  Thus, by the same argument above, adding in
+   * "ptr_tbl_size" cannot possibly overflow the "size" variable since
+   * it is a size_t object.  However, the extra pointers for auxv and
+   * the space for argv could cause an overflow.  The fact that we
+   * used stack to get here etc means that ptr_tbl_size could not have
+   * overflowed.
+   *
+   * NB: the underlying OS would have limited the amount of space used
+   * for argv and envv -- on linux, it is ARG_MAX, or 128KB -- and
+   * hence the overflow check is for obvious auditability rather than
+   * for correctness.
+   */
+  auxv_entries = 1;
+  if (0 != nap_child->user_entry_pt) {
+    auxv_entries++;
+  }
+  ptr_tbl_size = (((NACL_STACK_GETS_ARG ? 1 : 0) +
+                   (3 + argc + 1 + envc + 1 + auxv_entries * 2)) *
```

Comment:
these magic numbers should go away.

Reply:
good catch, will fix

Reply:
Turns out this is just the representation of pushing the argv, env, and auxv to the stack. The `STACK_GETS_ARG` is because on x86 it pushes an extra address because of SysV calling conventions. As esoteric as this is, there isn't really any better way to name these that I can think of.